feat: derive `Clone` for `mdns::Event`

drHuangMHT opened the pull request:
## Description

Derive trait `Clone` for `mdns::Event`. This makes cloning its contents without destroying type information possible.

<!--
Please write a summary of your changes and why you made them.
This section will appear as the commit message after merging.
Please craft it accordingly.
For a quick primer on good commit mesages, check out this blog post: https://cbea.ms/git-commit/

Please include any relevant issues in here, for example:

Related https://github.com/libp2p/rust-libp2p/issues/ABCD.
Fixes https://github.com/libp2p/rust-libp2p/issues/XYZ.
-->

Related #3593.

## Notes & open questions

CHANGELOG to be filled. Should we open a new minor version for the change?

<!--
Any notes, remarks or open questions you have to make about the PR which don't need to go into the final commit message.
-->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

drHuangMHT committed:
feat: make mdns::Event clone-able

Commit:
Merge branch 'libp2p:master' into master

thomaseizinger commented on protocols/mdns/src/behaviour.rs:
```diff
@@ -379,6 +380,7 @@ impl fmt::Debug for DiscoveredAddrsIter {
 }
 
 /// Iterator that produces the list of addresses that have expired.
+#[derive(Clone)]
```
Why do the iterators need to be `Clone` too?

drHuangMHT replied:
As the compiler suggests. Because the type is not a alias but a struct holding the iterator. All members should also be `Clone` when deriving `Clone` and a struct is not `Clone` by default.

thomaseizinger replied:
> All members should also be `Clone` when deriving `Clone` and a struct is not `Clone` by default.

Yes but why do these two structs need to be `Clone` in the first place? Is it not sufficient that the event they return is `Clone`? This means you can collect the iterator into a `Vec<mdns::Event>` and clone that vector as often as you want.

Iterators in Rust are typically single-use, I am not sure I want to make an API commitment that these ones can be cloned.

drHuangMHT replied:
I'm not comfortable with cloning iterators, too. So originally I changed the enum to hold `SmallVec` making it a breaking change. I'm experimenting with event listeners and type information makes delegating events to appropriate places easier. But if the event is single-use, it can't be send to multiple places with the type it comes from, causing inconsistency. I'm not so comfortable with that either.

drHuangMHT replied:
I don't think it's a good idea to construct data structures in places that shouldn't. I think it's similar to restrictions on `ListenerId`, even though enum variants are publicly available.

drHuangMHT replied:
Plus, the two structs have private field and there's no constructor for both of them currently, so there's actually no way to construct them outside the crate, let alone constructing `mdns::Event`.

thomaseizinger replied:
I am not sure I understand what you are getting at. With `mdns::Event` now being `Clone`, you can do the following with the iterator:

```rust
let peers = iter.collect::<Vec<_>>();
let peers2 = peers.clone();
```

What I am suggesting is to remove the `derive(Clone)` from the two iterator structs.

drHuangMHT replied:
As compiler suggests, you cannot derive `Clone` for `mdns::Event` unless the two structs are also `Clone`, because the variants hold them, and cloning the variant also clone the struct it holds. You can of course collect the results from the iterator with type `Vec<(PeerId,Multiaddr)>`, but you cannot put it back into the `Event` it comes with without access to the private field the struct holds. Although I can very well be designing a type as I want and implement `From<T>` and `Into<T>`, I still think it's a good idea to keep the original type information in case of confusion. It's also easier to trace where it comes from. If cloning in this way cannot be achieved, I'm reverting this PR. Everything started with the **unnecessary** iterator implementation, without which everyone is happy.

thomaseizinger replied:
Oh my god, I am a moron. Sorry for all this back and forth, I didn't look at the code properly. I thought `mdns::Event` was within those iterators and not the event holding the iterators :man_facepalming: 

> but you cannot put it back into the `Event`

That is true. Can you achieve your usecase without putting it back in? Typically, these events are only consumed by applications and translated into whatever data structure they need.

Sorry for going back on this but after having looked at everything properly now, I think having it `Clone` is unnecessary because you can simply extract the data and clone that.

drHuangMHT replied:
A vector-like structure can be cloned, or borrowed for better performance, and can create a iterator that borrows all its value. The vector is reusable in a sense while the iterator is not. Emit events that don't contain single-use elements feels more natural, especially when you have to explicitly convert it into a iterator before you emit the event, and that makes no sense.

drHuangMHT replied:
Again, collecting result is totally fine. But the type information is lost, making it harder to trace where it comes from. At least we can try to preserve the information for as long as possible before conversion is absolutely needed, and that's the whole point.

thomaseizinger replied:
> Again, collecting result is totally fine. But the type information is lost, making it harder to trace where it comes from. At least we can try to preserve the information for as long as possible before conversion is absolutely needed, and that's the whole point.

I guess that is fair. Happy to ship this. Thank you for your patience :)

drHuangMHT replied:
> > but you cannot put it back into the `Event`
> 
> That is true. Can you achieve your usecase without putting it back in? Typically, these events are only consumed by applications and translated into whatever data structure they need.

I am making an application and I can tell you very confidently that there's still some way to go before it gets converted. Plus, if all I want is just a vector, I need to collect the result manually, which is way more baffling than type `iter()` or `into_iter()` to get an iterator. I'll get upset in the former case while complete fine in the latter case because it's normal, it's part of life.

drHuangMHT committed:
docs: filling CHANGELOG for PR 3606

Unreleased patch version 0.43.1 bumped. Added entry for PR 3606: Deriving 'Clone' for 'mdns::Event'

drHuangMHT committed:
Merge branch 'master' of https://github.com/drHuangMHT/rust-libp2p

thomaseizinger commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+# 0.43.1 [unreleased]
+
+- Deriving `Clone` for `mdns::Event`. See [PR 3606].
+
```
```suggestion
- Deriving `Clone` for `mdns::Event`. See [PR 3606].

[PR 3606]: https://github.com/libp2p/rust-libp2p/pull/3606
```

thomaseizinger committed:
Update protocols/mdns/CHANGELOG.md

thomaseizinger commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+# 0.43.1 [unreleased]
+
+- Deriving `Clone` for `mdns::Event`. See [PR 3606].
```
```suggestion
- Derive `Clone` for `mdns::Event`. See [PR 3606].
```

thomaseizinger committed:
Update protocols/mdns/CHANGELOG.md

mergify[bot] committed:
Merge branch 'master' into master

mxinden commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+# 0.43.1 [unreleased]
```
This missed a version bump in `protocols/mdns/Cargo.toml`. I will make the bump with the upcoming release pull request.

mxinden replied:
Bump happening through https://github.com/libp2p/rust-libp2p/pull/3694.
